### ISSUE TITLE
Bug: the precision of sortition is not enough

### ISSUE THREAD

**Issue by Nanyan** [algorand/go-algorand#181]
### Description

The calculation precision  of the sortition (function `Select` in `data/committee/sortition/sortition.go`) is not enough.

For example: when `binomialP = 1e-6,  vrfOutput > "0xfffffffffffffcff000000000000000000000000000000000000000000000000"` , then the function `Select` will nearly always return a value equals the `binomialN`, that's the problem. 

If the sortition is under an exactly precision, with the above value and when `binomialN = 1e6` , then I think the correct `Select` result should be `18` rather than `1000000`.

### Consequence

With a bad precision, I think it will **weakness the effects of anti-sybil-attack**.

I make a test in `data/committee/credential_test.go`, after run several times, I got several result as follow (a line with red background means the `sybils` will get more votes than being honest)
![sybils-test-result](https://user-images.githubusercontent.com/1623062/61945152-d8182f00-afd1-11e9-9afb-e4ac5b6876a2.jpg)

The test code is as follow:  

```go
func TestSybilAttackRisk(t *testing.T) {
	n := 51
	// generate accounts
	genesis := make(map[basics.Address]*basics.AccountData)
	gen := rand.New(rand.NewSource(time.Now().Unix()))
	addrs := make([]basics.Address, n)
	secrets := make([]*crypto.SignatureSecrets, n)
	vrfSecrets := make([]*crypto.VrfPrivkey, n)
	otSecrets := make([]*crypto.OneTimeSignatureSecrets, n)
	proto := config.Consensus[protocol.ConsensusCurrentVersion]
	lookback := basics.Round(2*proto.SeedRefreshInterval + proto.SeedLookback + 1)
	total := basics.MicroAlgos{Raw:1e9}
	myStake := uint64(5e7)
	perSybilStake := myStake/uint64(n-1)
	for i := 0; i < n; i++ {
		addr, sigSec, vrfSec, otSec := newAccount(t, gen, lookback, 1000)
		addrs[i] = addr
		secrets[i] = sigSec
		vrfSecrets[i] = vrfSec
		otSecrets[i] = otSec

		short := addr
		genesis[short] = &basics.AccountData{
			Status:      basics.Online,
			MicroAlgos:  basics.MicroAlgos{Raw: perSybilStake},
			SelectionID: vrfSec.Pubkey(),
			VoteID:      otSec.OneTimeSignatureVerifier,
		}
	}
	genesis[addrs[0]].MicroAlgos.Raw = myStake

	rand.Seed(time.Now().Unix())
	var seed Seed
	rand.Read(seed[:])

	sel := AgreementSelector{
		Seed:   seed,
		Round:  basics.Round(1),
		Period: Period(1),
		Step:   Soft,
	}

	cnt, risk, rt, advantage, at := 1000, 0, 0, 0, 0
	for i := 0; i < cnt; i++ {
		addr0 := addrs[0]
		data0 := genesis[addr0]
		m := Membership{
			Record:     basics.BalanceRecord{Addr:addr0, AccountData:*data0},
			Selector:   sel,
			TotalMoney: total,
		}
		res,err := MakeCredential(vrfSecrets[0],sel).Verify(proto,m)
		var myWeight uint64
		if err == nil {
			myWeight = res.Weight
		}
		var sybilTotalWeight uint64
		for j := 1; j < n; j++ {
			addrj := addrs[j]
			dataj := genesis[addrj]
			m.Record = basics.BalanceRecord{Addr:addrj,AccountData:*dataj}
			res,err = MakeCredential(vrfSecrets[j],sel).Verify(proto,m)
			if err == nil {
				sybilTotalWeight += res.Weight
			}
		}
		at += int(myWeight)
		rt += int(sybilTotalWeight)
		if myWeight < sybilTotalWeight {
			risk++
			fmt.Printf("sybil  win: 0x%x  %d  %d\n", seed[:], myWeight, sybilTotalWeight)
		} else if myWeight > sybilTotalWeight {
			advantage++
			fmt.Printf("honest win: 0x%x  %d  %d\n", seed[:], myWeight, sybilTotalWeight)
		}
		rand.Read(seed[:])
		sel.Seed = seed
	}
	fmt.Printf("total: %d , honest win: %d, votes: %d, sybils win %d, votes: %d \n", cnt, advantage, at, risk, rt)
	if risk > advantage+advantage/20 {
		t.Errorf("potential sybil attack risk")
	}
}
```

**Comment by algoradam** [comment]
Thank you very much for reporting this! Just wanted to let you know that we're looking into it.

**Comment by derbear** [comment]
Hi @Nanyan,

- We changed the sortition code so `Select()` should now return
reasonable values when the VRF output is very high. In particular,
floating-point precision issues should now never result in getting
more votes than you would get if we had infinite precision. While this
doesn't actually increase the precision, it does fix the issue with
sybil-resistance. We've released a software update containing this
fix. (See https://github.com/algorand/go-algorand/commit/0ffcdca69a0673da6ec45400fd29e421a1d73232)

- We're considering increasing the precision beyond 64-bit floats, but
first we need to be sure that libboost's binomial cdf calculations are
both sufficiently accurate and consistent across platforms if we
increase the precision. We're also looking into other, simpler ways to
compute the sortition function and retain sybil-resistance.

Thanks again for looking into this sortition issue and reporting this 
bug! We really appreciate it. :)

- the Algorand security team
